src/hotspot/share/utilities/globalDefinitions.hpp

```diff
@@ -605,23 +606,24 @@ inline double percent_of(T numerator, T denominator) {
 //----------------------------------------------------------------------------------------------------
 // Special casts
 // Cast floats into same-size integers and vice-versa w/o changing bit-pattern
-typedef union {
-  jfloat f;
-  jint i;
-} FloatIntConv;
-
-typedef union {
-  jdouble d;
-  jlong l;
-  julong ul;
-} DoubleLongConv;
+template<typename T, typename F>
+inline T bit_cast(const F& from) noexcept { // replace with the real thing when we can use c++20
+  static_assert(sizeof(T) == sizeof(F), "must be of the same size");
+  static_assert(std::is_trivially_copyable<T>(), "output type must be trivially copyable");
+  static_assert(std::is_trivially_copyable<F>(), "input type must be trivially copyable");
+  static_assert(std::is_default_constructible<T>() &&
+                std::is_trivially_copy_assignable<T>(), "implementation limits");
+  T to;
+  memcpy(&to, &from, sizeof(T));
```

Comment:
During the review of JDK-8145096 it was found that some compilers produce
wretched code for these kinds of memcpy uses, even at fairly high optimization
levels.  (I don't know if we still care about those compilers.  Unfortunately I
don't remember which ones they were, other than gcc/clang/VS all being good.)

While using the so-called "union trick" is technically undefined behavior, it
is a technique that is known to be widely and well supported and produces good
code, at least for the cases where it is being used in HotSpot. In some cases,
such as gcc (and I think Visual Studio, though can't find a reference right
now), this behavior is documented.

Rather than adding a partial bit_cast (or moving it from elsewhere), we should
be using our existing PrimitiveConversions::cast
(metaprogramming/primitiveConversions.hpp). That has the small difficulty of a
circular include dependency with globalDefintions.hpp. That can be fixed by
moving the various jfoo_cast functions elsewhere (either to
primitiveConversions.hpp or to a new file; I might prefer the latter (along
with the Translate<jfloat/jdouble> specializations in primitiveConversions),
moving these relatively infrequently used utilities to their own dedicated
location). That also reduces the content of globalDefinitions.hpp, which IMO
is too much of a random dumping ground.

Reply:
From replies to what must by not be a GCC FAQ, type punning through a union like this is explicitly permitted:

```
inline jlong jlong_cast (jdouble x) {
  DoubleLongConv u;
  u.x = x;
  return u.l;
}
```
But this isn't:

```
inline jlong jlong_cast (jdouble x) {
  return ((DoubleLongConv*)&x)->l;
}
```
Not that it matters to HotSpot, because we use `-fno-strict-aliasing`.